lexer/lex.go

```diff
@@ -5,9 +5,164 @@ import (
 	"github.com/NeowayLabs/abad/token"
 )
 
-type (
-	Tokval struct {
-		Type  token.Type
-		Value utf16.Str
+type Tokval struct {
+	Type  token.Type
+	Value utf16.Str
+}
+
+func (t Tokval) Equal(other Tokval) bool {
+	return t.Type == other.Type && t.Value.Equal(other.Value)
+}
+
+// Lex will lex the given crappy JS code (utf16 yay) and provide a
+// stream of tokens as a result (the returned channel).
+//
+// The caller should iterate on the given channel until it is
+// closed indicating a EOF (or an error). Errors should be
+// handled by checking the type of the token.
+//
+// A goroutine will be started to lex the given code, if you
+// do not iterate the returned channel the goroutine will leak,
+// you MUST drain the provided channel.
+func Lex(code utf16.Str) <-chan Tokval {
+	tokens := make(chan Tokval)
+	
+	go func() {
+	
+		currentState := lexerInitialState(code)
+		
+		for currentState != nil {
+			token, newState := currentState()
+			if token != nil {
+				tokens <- *token
+			}
+			currentState = newState
+		}
+		
+		close(tokens)
+	}()
+
+	return tokens
+}
+
+type lexerState func() (*Tokval, lexerState)
+
+func lexerInitialState(code utf16.Str) lexerState {
+	return func() (*Tokval, lexerState) {
+		// TODO: handle empty input
+		
+		if len(code) == 0 {
+			return nil, nil
+		}
+		
+		if isNumber(code[0]) {
+			return nil, decimalOrHexadecimalState(code, 1)
+		}
+		
+		if isDot(code[0]) {
+			return nil, decimalState(code, 1)
+		}
+		
+		// TODO: Almost everything =)
+		return nil, nil
+	}
+}
+
+func decimalOrHexadecimalState(code utf16.Str, position uint) lexerState {
+	return func() (*Tokval, lexerState) {
+		if isEOF(code, position) {
+			return &Tokval{
+				Type: token.Decimal,
+				Value: code,
+			}, nil
+		}
+		
+		if isNumber(code[position]) || isDot(code[position]) {
```

Comment:
This is safe? I mean, you could be in the second byte of a multi-byte rune, then it could be a valid number but it's not.

Reply:
Already answered before =D (resuming I'm stupid)